src/charm.py

```diff
@@ -1,181 +1,99 @@
+                service_name="mlmd",
+                grpc_port=self._svc_grpc_port,
+                metadata_store_server_config_file=SQLITE_CONFIG_PROTO_DESTINATION,
+                files_to_push=[
+                    LazyContainerFileTemplate(
+                        destination_path=SQLITE_CONFIG_PROTO_DESTINATION,
+                        source_template=SQLITE_CONFIG_PROTO,
+                    )
                 ],
-            },
-            k8s_resources={
-                "kubernetesResources": {
-                    "services": [
-                        {
-                            "name": "metadata-grpc-service",
-                            "spec": {
-                                "selector": {"app.kubernetes.io/name": self.model.app.name},
-                                "ports": [
-                                    {
-                                        "name": "grpc-api",
-                                        "port": int(config["port"]),
-                                        "protocol": "TCP",
-                                        "targetPort": int(config["port"]),
-                                    },
-                                ],
-                            },
-                        }
-                    ]
-                }
-            },
+            ),
+            depends_on=[self.leadership_gate],
         )
-        self.model.unit.status = ActiveStatus()
-
-    def _send_info(self, interfaces):
-        if interfaces["grpc"]:
-            interfaces["grpc"].send_data(
-                {
-                    "service": "metadata-grpc-service",
-                    "port": self.model.config["port"],
-                }
-            )
-
-    def _check_leader(self):
-        if not self.unit.is_leader():
-            self.log.info("Not a leader, skipping set_pod_spec")
-            raise CheckFailed("", ActiveStatus)
-
-    def _get_interfaces(self):
-        try:
-            interfaces = get_interfaces(self)
-        except NoVersionsListed as err:
-            raise CheckFailed(err, WaitingStatus)
-        except NoCompatibleVersions as err:
-            raise CheckFailed(err, BlockedStatus)
-        return interfaces
-
-    def _check_image_details(self):
-        try:
-            image_details = self.image.fetch()
-        except OCIImageResourceError as e:
-            raise CheckFailed(f"{e.status.message}", e.status_type)
-        return image_details
-
-
-class CheckFailed(Exception):
-    """Raise this exception if one of the checks in main fails."""
-
-    def __init__(self, msg: str, status_type=None):
-        super().__init__()
-
-        self.msg = str(msg)
-        self.status_type = status_type
-        self.status = status_type(self.msg)
+
+        self.charm_reconciler.install_default_event_handlers()
+
+        grpc_port = ServicePort(int(self._svc_grpc_port), name="grpc-api")
+        self.service_patcher = KubernetesServicePatch(self, [grpc_port])
```

Comment:
Is it safe to expose the port AFTER calling `install_default_event_handlers`?
Now I'm not sure if it's a style choice we've done in the past, but usually we're doing that at the top of the init, so it's worth checking.

Reply:
The place where the k8s service patch is instantiated should not affect the result, afaik. What this library is doing is using lightkube to patch the `metadata-grpc-service Service`, which is a process that should not affect the result of other components (pebble, k8s, etc.). A quick way for testing this is to change the `port` config and checking the `Service`:

```
# With defaults
$ kubectl get svc -A | grep metadata-grpc-service
mlmd              metadata-grpc-service   ClusterIP   10.152.183.181   <none>        8080/TCP                 19m

# Change the config
$ juju config mlmd port=1010 # <---- any number just to see the port changes
$ kubectl get svc -A | grep metadata-grpc-service
mlmd              metadata-grpc-service   ClusterIP   10.152.183.181   <none>        1010/TCP                 31m
```

Reply:
I was concerned if it is needed for the pebble service to start since we are passing the port to the pebble component, but I think even if it's needed, we know that `pebble_ready` event handler will be executed after the init so you're right that it should not affect the result.
resolving this.